### ISSUE TITLE
Cannot retrieve package list behind proxy, SSL: CERTIFICATE_VERIFY_FAILED

### ISSUE THREAD

**Issue by lukrido** [wbond/package_control#1129]
Hello,

I am having an issue retrieving the list of packages to install in Sublime Text 3 build 3114.
I installed Package Control the manual way.
I am behind a corporate proxy and I have the following (redacted here) defined in my Package Control.sublime-settings

```
{
    "http_proxy": "proxy***com:***",
    "https_proxy": "proxy***com:***"
}
```

I have the same settings in my .cshrc as well and firefox can connect to the internet.

If I check https://packagecontrol.io/channel_v3.json from firefox (on the same machine), I have the following:
![image](https://cloud.githubusercontent.com/assets/5789068/17204600/ab24a39a-54a6-11e6-9d82-c155212d94a2.png)

![image](https://cloud.githubusercontent.com/assets/5789068/17204590/9e6b45b4-54a6-11e6-9446-fcce9938c34e.png)

However, when I log the connection attempt in Sublime Text, here's what I have:

```
Package Control: Download Debug
  URL: https://packagecontrol.io/submit?package=Package+Control&package_control_version=3.1.2&sublime_version=3114&version=3.1.2&sublime_platform=linux&operation=install
  Timeout: 30
  Resolved IP: 50.116.33.29
  Resolved IPv6: 2600:3c02::f03c:91ff:fedf:6883
Package Control: Urllib Debug Proxy
  http_proxy: proxy***.com:***
  https_proxy: proxy***.com:***
  proxy_username: 
  proxy_password: 
Package Control: Found system CA bundle at /etc/ssl/certs/ca-certificates.crt
Package Control: Urllib HTTPS Debug General
  Connecting to proxy***.com on port ***
Package Control: Urllib HTTPS Debug Write
  CONNECT packagecontrol.io:443 HTTP/1.1
  Proxy-Connection: Keep-Alive
  User-Agent: Package Control v3.1.2
  Host: packagecontrol.io:443
Package Control: Urllib HTTPS Debug Read
  HTTP/1.1 200 Connection established
Package Control: Urllib HTTPS Debug General
  Upgrading connection to SSL using CA certs file at /nfs/site/home/xxx/.config/sublime-text-3/Packages/User/Package Control.merged-ca-bundle
Package Control: Error submitting usage information. HTTP exception InvalidCertificateException (Host proxy***.com:*** returned an invalid certificate ([SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:548))) downloading https://packagecontrol.io/submit?package=Package+Control&package_control_version=3.1.2&sublime_version=3114&version=3.1.2&sublime_platform=linux&operation=install.
Package Control: Installing 1 missing dependency
Package Control: Fetching list of available packages
  Platform: linux-x64
  Sublime Text Version: 3114
  Package Control Version: 3.1.2
Package Control: Download Debug
  URL: https://packagecontrol.io/channel_v3.json
  Timeout: 30
  Resolved IP: 50.116.33.29
  Resolved IPv6: 2600:3c02::f03c:91ff:fedf:6883
Package Control: Urllib HTTPS Debug General
  Connecting to proxy***.com on port ***
Package Control: Urllib HTTPS Debug Write
  CONNECT packagecontrol.io:443 HTTP/1.1
  Proxy-Connection: Keep-Alive
  User-Agent: Package Control v3.1.2
  Host: packagecontrol.io:443
Package Control: Urllib HTTPS Debug Read
  HTTP/1.1 200 Connection established
Package Control: Urllib HTTPS Debug General
  Upgrading connection to SSL using CA certs file at /nfs/site/home/xxx/.config/sublime-text-3/Packages/User/Package Control.merged-ca-bundle
Package Control: Error downloading channel. HTTP exception InvalidCertificateException (Host proxy***.com:*** returned an invalid certificate ([SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:548))) downloading https://packagecontrol.io/channel_v3.json.
Package Control: The dependency specified, bz2, is not available
Package Control: Fetching list of available packages
  Platform: linux-x64
  Sublime Text Version: 3114
  Package Control Version: 3.1.2
Package Control: Download Debug
  URL: https://packagecontrol.io/channel_v3.json
  Timeout: 30
  Resolved IP: 50.116.33.29
  Resolved IPv6: 2600:3c02::f03c:91ff:fedf:6883
Package Control: Urllib HTTPS Debug General
  Connecting to proxy***.com on port ***
Package Control: Urllib HTTPS Debug Write
  CONNECT packagecontrol.io:443 HTTP/1.1
  Proxy-Connection: Keep-Alive
  User-Agent: Package Control v3.1.2
  Host: packagecontrol.io:443
Package Control: Urllib HTTPS Debug Read
  HTTP/1.1 200 Connection established
Package Control: Urllib HTTPS Debug General
  Upgrading connection to SSL using CA certs file at /nfs/site/home/xxx/.config/sublime-text-3/Packages/User/Package Control.merged-ca-bundle
Package Control: Error downloading channel. HTTP exception InvalidCertificateException (Host proxy***.com:*** returned an invalid certificate ([SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:548))) downloading https://packagecontrol.io/channel_v3.json.
Package Control: Fetching list of available packages
  Platform: linux-x64
  Sublime Text Version: 3114
  Package Control Version: 3.1.2
Package Control: Download Debug
  URL: https://packagecontrol.io/channel_v3.json
  Timeout: 30
  Resolved IP: 50.116.33.29
  Resolved IPv6: 2600:3c02::f03c:91ff:fedf:6883
Package Control: Urllib HTTPS Debug General
  Connecting to proxy***.com on port ***
Package Control: Urllib HTTPS Debug Write
  CONNECT packagecontrol.io:443 HTTP/1.1
  Proxy-Connection: Keep-Alive
  User-Agent: Package Control v3.1.2
  Host: packagecontrol.io:443
Package Control: Urllib HTTPS Debug Read
  HTTP/1.1 200 Connection established
Package Control: Urllib HTTPS Debug General
  Upgrading connection to SSL using CA certs file at /nfs/site/home/xxx/.config/sublime-text-3/Packages/User/Package Control.merged-ca-bundle
Package Control: Error downloading channel. HTTP exception InvalidCertificateException (Host proxy***.com:*** returned an invalid certificate ([SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:548))) downloading https://packagecontrol.io/channel_v3.json.
Package Control: No updated packages
```

It only uses the https_proxy variable.
My Package Control.merged-ca-bundle is empty.

I removed Package Control entirely and installed it again. Retrieving the list of packages still fails.

I asked IT and for now here's the statement:

> This does not appear to be a proxy problem, but rather a certificate problem between the application and the remote webserver, most likely due to an expired intermediary certificate.

**Comment by harshanarayana** [comment]
@lukrido 

Here's what I did to get around this messy problem while I was using SublimeText behind a proxy. (I am using Ubunut 16.04)


Get the "Package Control.sublime-package" from your current Installed location. In my case, it was inside `~/.config/sublime-text-3/Installed Packages`

Extract the Contents using Archive Manager > Navigate  to `Package Control>package_control>downloaders`

Open the curl_downloader.py file in your favorite editor. (I really hope it's sublime Text.)

Modify the Line #82 from 
`   command = [self.curl, '--connect-timeout', str_cls(int(timeout)), '-sSL',`
to 
`   command = [self.curl, '-k', '--connect-timeout', str_cls(int(timeout)), '-sSL',`

Save the file and Repackge the extracted Dir into the .sublime-package file again. 

Copy this one back to it's original location. 

Now, open the Custom User Settings for Package Control on your machine and add 
`
"downloader_precedence":
	{
		"linux":
		[
			"curl",
			"urllib",
			"wget"
		],
		"osx":
		[
			"urllib"
		],
		"windows":
		[
			"wininet"
		]
	},
`

This should fix your problem Temporarily. 

I KNOW I KNOW I KNOW. I am bypassing the secure option, but it gets my Sublime Text back to working.

**Comment by lukrido** [comment]
@wbond: do you have any update on this?
@harshanarayana: thanks for the suggestion, I will try that.

**Comment by wbond** [comment]
@lukrido I am certain there is no certificate issue with the webserver, I'm not sure why your IT department would suggest that. You can test it via https://www.ssllabs.com/ssltest/.

That said, there should definitely be certificates in `Package Control.merged-ca-bundle`. If not, that means Package Control has no CA certs to verify the connection with.

Is there anything inside of `/etc/ssl/certs/ca-certificates.crt`?

**Comment by schlamar** [comment]
@wbond This looks like an issue with Package Control to me. The logging message (Host proxy***.com:*** returned an invalid certificate) indicates that you are trying to validate the certificate against the proxy host and not against the target host of the HTTP request.

#1114 should be a duplicate of this issue.

Build 3114 startet to ship `_ssl` on Linux so I assume previously you didn't do HTTPS on Linux with urllib downloader? #1114 indicates that there isn't an issue with Build 3103.

**Comment by schlamar** [comment]
Oops I totally missed the point that the `Package Control.merged-ca-bundle` is empty.

So the validation of the certificate chain already fails which happens at the `ssl.wrap_socket` call. I assume there is something wrong reading the system CA certificates.

BTW I tested the urllib download with proxy on Windows and it is working fine!

**Comment by lukrido** [comment]
@schlamar when you say

> So the validation of the certificate chain already fails which happens at the ssl.wrap_socket call. I assume there is something wrong reading the system CA certificates.

do you mean it is something on my side or on Package Control side?

**Comment by wbond** [comment]
Check and see if Package Control 3.3.0 resolves the issue. I believe this may be tied to SNI support.

**Comment by lukrido** [comment]
Hello @wbond, I didn't see the update last time...

I downloaded Package Control 3.3.0 and I have the same issue.
Here's my log:
`Package Control: Skipping automatic upgrade, last run at 2017-12-20 12:04:19, next run at 2017-12-20 13:04:19 or after
Package Control: Fetching list of available packages and dependencies
  Platform: linux-x64
  Sublime Text Version: 3114
  Package Control Version: 3.3.0
Package Control: Download Debug
  URL: https://packagecontrol.io/channel_v3.json
  Timeout: 30
  Resolved IP: 50.116.34.243
  Resolved IPv6: 2600:3c02:e000:42::1
Package Control: Urllib Debug Proxy
  http_proxy: proxy***.com:***
  https_proxy: proxy***.com:***
  proxy_username: 
  proxy_password: 
Package Control: Found system CA bundle at /etc/ssl/certs/ca-certificates.crt (101247 bytes)
Package Control: Regenerated the merged CA bundle from the system and user CA bundles
Package Control: Urllib HTTPS Debug General
  Connecting to proxy***.com on port ***
Package Control: Urllib HTTPS Debug Write
  CONNECT packagecontrol.io:443 HTTP/1.1
  User-Agent: Package Control v3.3.0
  Host: packagecontrol.io:443
  Proxy-Connection: Keep-Alive
Package Control: Urllib HTTPS Debug Read
  HTTP/1.1 200 Connection established
Package Control: Urllib HTTPS Debug General
  Upgrading connection to SSL using CA certs file at /nfs/site/home/xxx/.config/sublime-text-3/Packages/User/Package Control.merged-ca-bundle
  Using hostname "packagecontrol.io" for TLS SNI extension
Package Control: Error downloading channel. HTTP exception InvalidCertificateException (Host proxy***.com:*** returned an invalid certificate ([SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:548))) downloading https://packagecontrol.io/channel_v3.json.
`
